File: privacystreams-app/src/main/java/io/github/privacystreams/app/PStreamCollectService.java
package io.github.privacystreams.app;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import io.github.privacystreams.app.db.PSLocationDBHelper;
import io.github.privacystreams.app.db.PStreamDBHelper;

/**
 * The PrivacyStreams always-on service for collecting historic data.
 */

public class PStreamCollectService extends Service {
    private static final int ONGOING_NOTIFICATION_ID = 1;

    List<PStreamDBHelper> dbHelpers = new ArrayList<>();

    @Override
    public void onCreate() {
        this.dbHelpers.add(new PSLocationDBHelper(this));
        for (PStreamDBHelper dbHelper : this.dbHelpers) {
            dbHelper.startCollecting();
        }

        Intent notificationIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, notificationIntent, 0);

        Notification notification = new Notification.Builder(this)
                .setContentTitle(getText(R.string.collect_notification_title))
                .setContentText(getText(R.string.collect_notification_text))
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentIntent(pendingIntent)
                .setTicker(getText(R.string.collect_notification_title))
                .build();

        startForeground(ONGOING_NOTIFICATION_ID, notification);
    }

    @Override
    public void onDestroy() {
        for (PStreamDBHelper dbHelper : this.dbHelpers) {
            dbHelper.stopCollecting();
        }

        stopForeground(true);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }
}

File: privacystreams-app/src/main/java/io/github/privacystreams/app/PStreamCollectService.kt
package io.github.privacystreams.app

import android.app.Notification
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.IBinder

import io.github.privacystreams.app.db.PStreamDBHelper

/**
 * The PrivacyStreams always-on service for collecting historic data.
 */

class PStreamCollectService : Service() {

    internal val dbHelper = PStreamDBHelper.getInstance(this)
    internal val dbTables = dbHelper.tables

    override fun onCreate() {
        dbTables.map { it.startCollecting() }

        val notificationIntent = Intent(this, NavActivity::class.java)
        val bundle = Bundle()
        bundle.putInt(NavActivity.NAV_ID_KEY, R.id.nav_data)
        notificationIntent.putExtras(bundle)
        val pendingIntent = PendingIntent.getActivity(this, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT)

        val notification = Notification.Builder(this)
                .setContentTitle(getText(R.string.collect_notification_title))
                .setContentText(getText(R.string.collect_notification_text))
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentIntent(pendingIntent)
                .setTicker(getText(R.string.collect_notification_title))
                .build()

        startForeground(ONGOING_NOTIFICATION_ID, notification)
    }

    override fun onDestroy() {
        dbTables.map { it.stopCollecting() }
        stopForeground(true)
        dbHelper.close()
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return Service.START_STICKY
    }

    companion object {
        private val ONGOING_NOTIFICATION_ID = 1
    }
}
